mesos-client/src/main/scala/com/mesosphere/mesos/client/Session.scala

```diff
@@ -0,0 +1,220 @@
+    }
+
+  /** @return The connection pool for this session. */
+  def connectionPool(implicit system: ActorSystem): Flow[HttpRequest, Try[HttpResponse], NotUsed] = {
+    // Disable pipelining.
+    val poolSettings = ConnectionPoolSettings("").withMaxConnections(1).withPipeliningLimit(1)
+    if (isSecured) {
+      Flow[HttpRequest]
+        .map(_ -> NotUsed)
+        .via(Http().cachedHostConnectionPoolHttps(host = url.getHost, port = port, settings = poolSettings))
+        .map(_._1)
+    } else {
+      Flow[HttpRequest]
+        .map(_ -> NotUsed)
+        .via(Http().cachedHostConnectionPool(host = url.getHost, port = port, settings = poolSettings))
+        .map(_._1)
+    }
+  }
+}
+
+object SessionActor {
+  def props(
+      credentialsProvider: CredentialsProvider,
+      requestFactory: (Array[Byte], Option[HttpCredentials]) => HttpRequest): Props = {
+    Props(new SessionActor(credentialsProvider, requestFactory))
+  }
+}
+
+class SessionActor(
+    credentialsProvider: CredentialsProvider,
+    requestFactory: (Array[Byte], Option[HttpCredentials]) => HttpRequest)
+    extends Actor
+    with Stash
+    with StrictLogging {
+  import akka.pattern.pipe
+  import scala.concurrent.ExecutionContext.Implicits.global
+
+  case class Response(originalCall: Array[Byte], originalSender: ActorRef, response: HttpResponse)
+
+  override def preStart(): Unit = {
+    super.preStart()
+    credentialsProvider.nextToken().pipeTo(self)
+  }
+
+  override def receive: Receive = initializing
+
+  def initializing: Receive = {
+    // TODO: Handle nextToken errors.
+    case credentials: HttpCredentials =>
+      logger.info("Retrieved token")
+      context.become(initialized(credentials))
+      unstashAll()
+    case _ => stash()
+  }
+
+  def initialized(credentials: HttpCredentials): Receive = {
+    case call: Array[Byte] =>
+      val request = requestFactory(call, Some(credentials))
+      val originalSender = sender()
+      logger.info("Processing next call.")
+      Http()(context.system)
+        .singleRequest(request)
+        .onComplete {
+          case Success(response) =>
+            logger.info(s"HTTP response: ${response.status}")
+            self ! Response(call, originalSender, response)
+          case Failure(ex) =>
+            logger.error("HTTP request failed", ex)
+            // Fail stream.
+            originalSender ! Status.Failure(ex)
+        }
+    case Response(originalCall, originalSender, response) =>
+      logger.info(s"Call replied with ${response.status}")
+      if (response.status == StatusCodes.Unauthorized) {
+        logger.info("Refreshing token")
+
+        context.become(initializing)
+        credentialsProvider.nextToken().pipeTo(self)
+
+        // Queue current call again.
+        self ! originalCall
+      } else {
+        originalSender ! response
+      }
+  }
+}
+
+case class SessionFlow(
```

Comment:
I'm not entirely convinced that a custom graph stage is needed here; can we perhaps go over this tomorrow together?

Reply:
That's what the walkthrough is for. I'd use the actor below. Remember we have to keep the session token state somewhere and refresh it if a request returns an `HTTP 401 Unauthorized`. After some research I'm uncertain that a stream could handle this easier than a simple actor. However, I'm open to suggestions.